typescript/rebalance_bot/run.ts

```diff
@@ -0,0 +1,191 @@
+  }
+
+  async start(): Promise<void> {
+    if (!this.rebalanceManager) {
+      throw new Error("Bot not initialized. Call initialize() first.");
+    }
+
+    this.isRunning = true;
+    logger.info(
+      `Starting bot loop with ${this.config.policy.loopIntervalSec}s interval`,
+    );
+
+    while (this.isRunning) {
+      try {
+        await this.runCycle();
+      } catch (error) {
+        logger.error("Bot cycle failed:", error);
+      }
+
+      if (this.isRunning) {
+        logger.debug(
+          `Sleeping for ${this.config.policy.loopIntervalSec} seconds`,
+        );
+        await this.sleep(this.config.policy.loopIntervalSec * 1000);
+      }
+    }
+
+    logger.info("Bot stopped");
+  }
+
+  stop(): void {
+    logger.info("Stopping bot...");
+    this.isRunning = false;
+  }
+
+  private async runCycle(): Promise<void> {
+    logger.debug("Starting rebalance cycle");
+
+    // Check if we should skip this cycle due to recent ignore
+    if (this.shouldIgnoreCycle()) {
+      logger.debug("Skipping cycle due to recent ignore entry");
+      return;
+    }
+
+    try {
+      await this.rebalanceManager!.executeRebalance();
+    } catch (error) {
+      logger.error("Rebalance execution failed:", error);
+
+      // Add ignore entry to prevent spam
+      this.addIgnoreEntry("execution_failed");
+    }
+  }
+
+  private shouldIgnoreCycle(): boolean {
+    const ignoreData = this.ignoreCache.load<{ [key: string]: IgnoreEntry }>();
+    if (!ignoreData) {
+      return false;
+    }
+
+    const coreAddress = this.config.contracts.dloopCore;
+    const entry = ignoreData[coreAddress];
+
+    if (!entry) {
+      return false;
+    }
+
+    // Ignore for configured duration after last entry
+    const ignoreUntil = entry.timestamp + IGNORE_DURATION_MS;
+    const now = Date.now();
+
+    if (now < ignoreUntil) {
+      logger.debug(
+        `Ignoring until ${new Date(ignoreUntil).toISOString()}, reason: ${entry.reason}`,
+      );
+      return true;
+    }
+
+    return false;
+  }
+
+  private addIgnoreEntry(reason: string): void {
+    const ignoreData =
+      this.ignoreCache.load<{ [key: string]: IgnoreEntry }>() || {};
+    const coreAddress = this.config.contracts.dloopCore;
+
+    ignoreData[coreAddress] = {
+      timestamp: Date.now(),
+      reason,
+    };
+
+    this.ignoreCache.save(ignoreData);
+    logger.debug(`Added ignore entry for ${coreAddress}: ${reason}`);
+  }
+
+  private sleep(ms: number): Promise<void> {
+    return new Promise((resolve) => setTimeout(resolve, ms));
+  }
+}
+
+// Main entry point
+async function main(): Promise<void> {
+  const bot = new RebalanceBotRunner();
+
+  // Handle graceful shutdown
+  process.on("SIGINT", () => {
+    logger.info("Received SIGINT, shutting down gracefully...");
+    bot.stop();
+  });
+
+  process.on("SIGTERM", () => {
+    logger.info("Received SIGTERM, shutting down gracefully...");
+    bot.stop();
+  });
+
+  // Handle unhandled promise rejections
```

Comment:
Excellent addition of global error handlers! The differentiation between connection errors (which stop the bot) and other unhandled rejections (which allow continuation) is a good design decision. This will prevent the bot from crashing unexpectedly in production.